### PR TITLE
Fixing some issues and adding some features.

### PR DESCRIPTION
**Pull request by simonfqy**
The `metric/__init__.py` script has a line which would cause errors, so I fixed it. I also assume binary classification when using thresholds.

The `BatchNorm()` function in `layers.py` does not incorporate training flag in the `tf.layers.batch_normalization()` function it calls, so it could not be performing as expected. Previously I used the `BatchNormalization()` function in the same script, after adding this flag to `BatchNorm()` function and using it instead, I found out that once this training flag is included, there could be some `NAN` arising from the `WeaveGather` class. After examining this issue with `tf_debug`, I found this problem is caused by division by zero. Due to the erratic crashing behavior of `tf_debug` on my machine, I am not able to figure out why this problem would not happen when using `BatchNormalization`.

The `self.epsilon` of `WeaveGather` class does not do anything, so I used it to solve the division by zero problems mentioned above.

### CODE REVIEW

**Comment by simonfqy** [comment]
I don't know why my Travis CI build failed. This is my first time committing to this kind of public projects. Any explanation is greatly appreciated!

**Comment by rbharath** [comment]
@simonfqy Thanks for the contribution! This PR has a number of unrelated changes. That makes it hard for us to gauge whether the changes are correct or not.

For a first PR, I'd recommend picking one of the changes, and then adding unit tests that check whether the feature has been implemented correctly. That way, we can understand the effect of the change and gauge whether it works as desired.

**Comment by simonfqy** [comment]
@rbharath Thanks for your reply. I will do so in this week. Hope that you keep the pull request open for me to update it.

**Comment by rbharath** [comment]
@simonfqy I'm going to close this old PR. Feel free to re-open a fresh PR if you're still interested in getting this in!
